### PR TITLE
Shrink `correction` buffer when appropriate

### PR DESCRIPTION
**Pull request by frankmcsherry**
The `correction` buffer in `persist_sink` starts large as it diffs the initial snapshots, but in steady state contains substantially fewer updates. However, we never shrink it from its peak size. This PR introduces a `shrink_to_fit` call when the capacity is more than 4x the actual length of the corrections, which we do to ensure that the allocation does not then immediately double if a single correction is added.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a
  companion cloud PR to account for those changes that is tagged with
  the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Comment by frankmcsherry** [comment]
This resulted in a 10GB reduction in memory footprint for me with one materialized view (containing 1B elements).

File: src/compute/src/sink/persist_sink.rs [hunk 1]
```diff
@@ -745,6 +745,9 @@ where
                     // spend a lot of time "consolidating" the same updates
                     // over and over again, with no changes.
                     consolidate_updates(&mut correction);
+                    if correction.len() < correction.capacity() / 4 {
```
Comment (danhhz) [inline src/compute/src/sink/persist_sink.rs:748]: copy the PR description in here as a comment, please!

**Review by danhhz — APPROVED** [review]
LGTM modulo one comment nit, thanks!!

File: src/compute/src/sink/persist_sink.rs (hunk 1 above)
Comment (frankmcsherry) [reply src/compute/src/sink/persist_sink.rs:759]: Fixed up in the subsequent commit.

**Review by antiguru — APPROVED** [review]
LGTM!

**Comment by frankmcsherry** [comment]
Thanks for the the reviews!

**Pull request merged**